girder/models/file.py

```diff
@@ -93,19 +94,42 @@ def download(self, file, offset=0, headers=True, endByte=None,
         :type contentDisposition: str or None
         :type extraParameters: str or None
         """
+        if offset is 0:
+            events.trigger('model.file.download.before', info={'file': file})
+        else:
+            events.trigger('model.file.download.during', info={'file': file})
         if file.get('assetstoreId'):
-            return self.getAssetstoreAdapter(file).downloadFile(
-                file, offset=offset, headers=headers, endByte=endByte,
-                contentDisposition=contentDisposition,
-                extraParameters=extraParameters)
+            try:
+                fileDownload = self.getAssetstoreAdapter(file).downloadFile(
+                    file, offset=offset, headers=headers, endByte=endByte,
+                    contentDisposition=contentDisposition,
+                    extraParameters=extraParameters)
+                if inspect.isgeneratorfunction(fileDownload) or inspect.isgenerator(fileDownload):
+                    def downloadGenerator():
+                        for data in fileDownload():
+                            yield data
+                        if endByte is None or endByte >= file['size']:
+                            events.trigger('model.file.download.after',
+                                           info={'file': file, 'redirect': False})
+                    return downloadGenerator
+                else:
+                    # With our current assetstores (June 2017), this case should not be reached.
```

Comment:
Maybe this should trigger an `'model.file.download.after'` event, just for safety in the future?